commons/src/test/java/io/aiven/kafka/tieredstorage/commons/storage/filesystem/FileSystemStorageTest.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.commons.storage.filesystem;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class FileSystemStorageTest {
+
+    @Test
+    void testRootCannotBeAFile(@TempDir final Path root) throws IOException {
+        final Path wrongRoot = root.resolve("file_instead");
+        Files.writeString(wrongRoot, "Wrong root");
+
+        assertThrows(IllegalArgumentException.class, () -> new FileSystemStorage(wrongRoot, true));
+    }
+
+    @Test
+    void testRootCannotBeNonWritableDirectory(@TempDir final Path root) throws IOException {
+        final Path nonWritableDir = root.resolve("non_writable");
+        final FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(
+            Set.of(
+                PosixFilePermission.OWNER_READ,
+                PosixFilePermission.GROUP_READ,
+                PosixFilePermission.OTHERS_READ
+            ));
+        Files.createDirectory(nonWritableDir, fileAttributes);
+
+        assertThrows(IllegalArgumentException.class, () -> new FileSystemStorage(nonWritableDir, true));
+    }
+
+    @Test
+    void testUploadANewFile(@TempDir final Path root) throws IOException {
+        final FileSystemStorage storage = new FileSystemStorage(root, false);
+        final String content = "content";
+        final String key = "key";
+        storage.upload(new ByteArrayInputStream(content.getBytes()), key);
+
+        assertEquals(content, Files.readString(root.resolve(key)));
+    }
+
+    @Test
+    void testUploadFailsWhenFileExists(@TempDir final Path root) throws IOException {
+        final String key = "key";
```

Comment:
Would it make sense to create constant for this dummy key instead of creating a variable in almost every test method?

Reply:
Sure, creating a constant for this representing the topic/partition/segment structure to be used.